File: app/src/main/java/com/saint/struct/data/PagingOldDataSource.java
package com.saint.struct.data;

import android.util.Log;

import androidx.annotation.NonNull;
import androidx.paging.PageKeyedDataSource;

import com.saint.struct.bean.WanAndroidBean;
import com.saint.struct.bean.WanListBean;
import com.saint.struct.network.HttpManager;
import com.saint.struct.network.service.ConnectService;

import java.util.ArrayList;

import retrofit2.Call;
import retrofit2.Callback;
import retrofit2.Response;

public class PagingOldDataSource extends PageKeyedDataSource<String, WanListBean> {
    private static final String TAG = PagingOldDataSource.class.getSimpleName();
    private int mPage = 0;

    @Override
    public void loadInitial(@NonNull LoadInitialParams<String> params, @NonNull LoadInitialCallback<String, WanListBean> callback) {
        ConnectService api = HttpManager.getInstance().getService();
        Call<WanAndroidBean> call = api.getArticleList(mPage, 60);
        call.enqueue(new Callback<WanAndroidBean>() {
            @Override
            public void onResponse(Call<WanAndroidBean> call, Response<WanAndroidBean> response) {
                Log.e(TAG, "--onResponse-->" + response.body().data.datas.size());
                if (response.isSuccessful() && response.code() == 200) {
                    ArrayList<WanListBean> data = response.body().data.datas;
                    callback.onResult(data, "before", "after");
                }

            }

            @Override
            public void onFailure(Call<WanAndroidBean> call, Throwable t) {
                Log.e(TAG, "--onFailure-->" + t.getMessage());
            }
        });
    }

    @Override
    public void loadBefore(@NonNull LoadParams<String> params, @NonNull LoadCallback<String, WanListBean> callback) {
        Log.i(TAG, "--loadBefore-->" + params.key);
    }

    @Override
    public void loadAfter(@NonNull LoadParams<String> params, @NonNull LoadCallback<String, WanListBean> callback) {
        mPage++;
        ConnectService api = HttpManager.getInstance().getService();
        Call<WanAndroidBean> call = api.getArticleList(mPage, 60);
        call.enqueue(new Callback<WanAndroidBean>() {
            @Override
            public void onResponse(Call<WanAndroidBean> call, Response<WanAndroidBean> response) {
                if (response.isSuccessful() && response.code() == 200) {
                    ArrayList<WanListBean> data = response.body().data.datas;
                    callback.onResult(data, params.key);
                }

            }

            @Override
            public void onFailure(Call<WanAndroidBean> call, Throwable t) {
                Log.e(TAG, "--onFailure-->" + t.getMessage());
            }
        });
    }

}

File: app/src/main/java/com/saint/struct/data/PagingOldDataSource.kt
package com.saint.struct.data

import android.util.Log
import androidx.paging.PageKeyedDataSource
import com.saint.struct.bean.WanAndroidBean
import com.saint.struct.bean.WanListBean
import com.saint.struct.data.PagingOldDataSource
import com.saint.struct.network.HttpManager
import retrofit2.Call
import retrofit2.Callback
import retrofit2.Response

class PagingOldDataSource : PageKeyedDataSource<String, WanListBean>() {
    private var mPage = 0
    override fun loadInitial(params: LoadInitialParams<String>, callback: LoadInitialCallback<String, WanListBean>) {
        val api = HttpManager.getInstance().service
        val call = api.getArticleList(mPage, 60)
        call?.enqueue(object : Callback<WanAndroidBean> {
            override fun onResponse(call: Call<WanAndroidBean>, response: Response<WanAndroidBean>) {
                Log.e(TAG, "--onResponse-->" + response.body()!!.data!!.datas.size)
                if (response.isSuccessful && response.code() == 200) {
                    val data: ArrayList<WanListBean> = response.body()!!.data!!.datas
                    callback.onResult(data, "before", "after")
                }
            }

            override fun onFailure(call: Call<WanAndroidBean>, t: Throwable) {
                Log.e(TAG, "--onFailure-->" + t.message)
            }
        })
    }

    override fun loadBefore(params: LoadParams<String>, callback: LoadCallback<String, WanListBean>) {
        Log.i(TAG, "--loadBefore-->" + params.key)
    }

    override fun loadAfter(params: LoadParams<String>, callback: LoadCallback<String, WanListBean>) {
        mPage++
        val api = HttpManager.getInstance().service
        val call = api.getArticleList(mPage, 60)
        call?.enqueue(object : Callback<WanAndroidBean> {
            override fun onResponse(call: Call<WanAndroidBean>, response: Response<WanAndroidBean>) {
                if (response.isSuccessful && response.code() == 200) {
                    val data: ArrayList<WanListBean> = response.body()!!.data!!.datas
                    callback.onResult(data, params.key)
                }
            }

            override fun onFailure(call: Call<WanAndroidBean>, t: Throwable) {
                Log.e(TAG, "--onFailure-->" + t.message)
            }
        })
    }

    companion object {
        private val TAG = PagingOldDataSource::class.java.simpleName
    }
}